extensions/wikia/UserRenameTool/SpecialRenameuser_body.php

```diff
@@ -32,99 +31,179 @@ public function execute( $par ) {
-		$token = $this->getUser()->getEditToken();
-		$notifyRenamed = $request->getBool( 'notify_renamed', false );
-		$confirmAction = false;
+		// TODO: remove after QA tests
+		if ( self::shouldUnlock() ) {
+			$requestorUser->setGlobalFlag( 'wasRenamed', false );
+			$requestorUser->saveSettings();
+		}
+		// TODO: end
 
-		if ( $request->wasPosted() && $request->getInt( 'confirmaction' ) ) {
-			$confirmAction = true;
+		if ( \RenameUserHelper::canUserChangeUsername( $requestorUser ) ) {
+			$this->renderForm( $requestorUser, $username );
+		} else {
+			$this->renderDisallow();
 		}
 
-		$warnings = [];
-		$errors = [];
-		$info = [];
-
-		if ( $this->validRenameRequest() ) {
-			$process = new RenameUserProcess( $oldUsername, $newUsername, $confirmAction, $reason, $notifyRenamed );
-			$status = $process->run();
-			$warnings = $process->getWarnings();
-			$errors = $process->getErrors();
-			if ( $status ) {
-				$info[] = $this->msg( 'userrenametool-info-in-progress' )->inContentLanguage()->escaped();
-			}
+		return;
+	}
+
+	public static function validateData( array $data, User $user, bool $selfRename = true ) {
+		$errorList = [];
+		$newUsername = $data['newUsername'];
+
+		if ( $data['token'] === '' ) {
+			$errorList[] = 'userrenametool-error-token-not-exists';
 		}
 
-		$showConfirm = ( empty( $errors ) && empty( $info ) );
+		if ( $newUsername === '' ) {
+			$errorList[] = 'userrenametool-error-no-username';
+		}
 
-		// note: errors and info beyond this point are non-blocking
+		if ( $newUsername !== $data['newUsernameRepeat' ] ) {
+			$errorList[] = 'userrenametool-error-not-repeated-correctly';
+		}
 
-		if ( !empty( $oldUsername ) ) {
-			$oldUser = User::newFromName( $oldUsername );
-			if ( $oldUser->getGlobalFlag( 'requested-rename', 0 ) ) {
-				$info[] = $this->msg( 'userrenametool-requested-rename', $oldUsername )->escaped();
-			} else {
-				$errors[] = $this->msg( 'userrenametool-did-not-request-rename', $oldUsername )->escaped();
-			}
-			if ( $oldUser->getGlobalFlag( 'wasRenamed', 0 ) ) {
-				$errors[] = $this->msg( 'userrenametool-previously-renamed', $oldUsername )->escaped();
+		if ( !\User::isValidUserName( $newUsername ) ) {
+			$errorList[] = 'userrenametool-error-non-alphanumeric';
+		}
+
+		if ( strlen( $newUsername ) > \RenameUserHelper::MAX_USERNAME_LENGTH ) {
+			$errorList[] = 'userrenametool-error-too-long';
+		}
+
+		if ( $data['understandConsequences'] !== 'true' ) {
+			$errorList[] = 'userrenametool-error-consequences';
+		}
+
+		if ( !$user->matchEditToken( $data['token'] ) ) {
+			$errorList[] = 'userrenametool-error-token-not-match';
+		}
+
+		if ( $selfRename && !$user->checkPassword( $data['password'] )->success() ) {
+			$errorList[] = 'userrenametool-error-password-not-match';
+		}
+
+		if ( !\RenameUserHelper::canUserChangeUsername( $user ) ) {
+			$errorList[] = 'userrenametool-error-alreadyrenamed';
+		}
+
+		return $errorList;
+	}
+
+	// TODO: remove after QA tests
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1135)

[project-id]: wikia:app